File: fs/Practice2/DynamicProgramming/Bit/RevengeOfTravelingSalesmanProblem.fs
// https://atcoder.jp/contests/s8pc-1/tasks/s8pc_1_g

(fun () -> stdin.ReadLine () |> fun x -> x.Split ' ')
|> fun readLine ->
  readLine ()
  |> Array.map int
  |> fun arr -> (arr.[0], arr.[1])
  |> fun (n, m) ->
    (
      List.init m (fun _ ->
        readLine ()
        |> fun arr -> (int arr.[0] - 1, int arr.[1] - 1, int64 arr.[2], int64 arr.[3])
      )
      |> List.map (fun (s, t, d, time) -> if s < t then (s, t, d, time) else (t, s, d, time)),
      Array.init (1 <<< n) (fun _ -> Array.create n (System.Int64.MaxValue, 0L))
    )
    |> fun (graph, memo) ->
      (memo.[1].[0] <- (0L, 1L))
      |> fun () ->
        [
          seq {
            for vx in 1..2..(1 <<< n) - 1 do
            for (s, t, d, time) in graph do
            if vx >>> s &&& 1 = 1 && vx >>> t &&& 1 = 1
            then
              yield!
                [
                  (vx, t, vx ^^^ (1 <<< t), s, d, time);
                  (vx, s, vx ^^^ (1 <<< s), t, d, time);
                ]
          };
          seq {
            for (s, t, d, time) in graph do
            if s = 0
            then yield (0, 0, (1 <<< n) - 1, t, d, time)
          };
        ]
        |> Seq.concat
        |> Seq.iter (
          fun (i, j, i2, j2, d, time) ->
            memo.[i2].[j2]
            |> fun (preV, preC) ->
              if preV <> System.Int64.MaxValue && preV + d <= time
              then (
                (memo.[i].[j], (preV + d))
                |> fun ((accV, accC), curV) ->
                  match i with
                  | _ when curV < accV -> (curV, preC)
                  | _ when curV = accV -> (accV, accC + preC)
                  | _ -> (accV, accC)
                |> fun x -> memo.[i].[j] <- x
              )
        )
      |> fun _ -> memo.[0].[0]
|> fun (v, c) -> if c = 0L then "IMPOSSIBLE" else sprintf "%d %d" v c
|> printfn "%s"

// input:
// 3 3
// 1 2 1 6
// 2 3 2 6
// 3 1 3 6

// memo:
// [|
//   [| (6, 2); (-, 0); (-, 0) |] // 0: 000
//   [| (0, 1); (-, 0); (-, 0) |] // 1: 001
//   [| (-, 0); (-, 0); (-, 0) |] // 2: 010
//   [| (-, 0); (1, 1); (-, 0) |] // 3: 011
//   [| (-, 0); (-, 0); (-, 0) |] // 4: 100
//   [| (-, 0); (-, 0); (3, 1) |] // 5: 101
//   [| (-, 0); (-, 0); (-, 0) |] // 6: 110
//   [| (-, 0); (5, 1); (3, 1) |] // 7: 111
// |]

// output:
// 6 2

// https://atcoder.jp/contests/s8pc-1/submissions/20735572

File: scala/Practice2/DynamicProgramming/Bit/RevengeOfTravelingSalesmanProblem.scala
// https://atcoder.jp/contests/s8pc-1/tasks/s8pc_1_g

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    (() => io.StdIn.readLine.split(' '))
    |> (readLine => (
      readLine().map(_.toInt)
      |> (arr => (arr(0), arr(1)))
      |> { case (n, m) => (
        (
          List.range(0, m)
            .map(_x => (readLine() |> (arr => (arr(0).toInt - 1, arr(1).toInt - 1, arr(2).toLong, arr(3).toLong))))
            .map { case (s, t, d, time) => if (s < t) (s, t, d, time) else (t, s, d, time) },
          Array.range(0, 1 << n)
            .map(_x => Array.fill[Tuple2[Long, Long]](n)(Long.MaxValue, 0L))
        )
        |> { case (graph, memo) => (
          (memo(1)(0) = (0L, 1L))
          |> (_x =>
            Stream.empty
              .append(
                for (
                  vx <- 1 until (1 << n) by 2;
                  (s, t, d, time) <- graph if (vx >> s & 1) == 1 && (vx >> t & 1) == 1;
                  x <- Stream(
                    (vx, t, d, time, vx ^ (1 << t), s),
                    (vx, s, d, time, vx ^ (1 << s), t)
                  )
                )
                yield x
              )
              .append(
                for (
                  (s, t, d, time) <- graph if s == 0;
                  x <- Stream(
                    (0, 0, d, time, (1 << n) - 1, t)
                  )
                )
                yield x
              )
              .foreach { case (i, j, d, time, i2, j2) => (
                memo(i2)(j2)
                |> { case (prevCost, prevCount) =>
                  if (prevCost != Long.MaxValue && prevCost + d <= time) (
                    (memo(i)(j), prevCost + d)
                    |> { case ((accCost, accCount), curCost) =>
                      i match {
                        case _x if curCost < accCost => (curCost, prevCount)
                        case _x if curCost == accCost => (accCost, accCount + prevCount)
                        case _x => (accCost, accCount)
                      }
                    }
                    |> (v => memo(i)(j) = v)
                  )
                }
              )}
          )
          |> (_x => memo(0)(0))
        )}
      )}
    ))
    |> { case (cost, count) => if (count == 0L) "IMPOSSIBLE" else s"${cost} ${count}" }
    |> println
  )
}
